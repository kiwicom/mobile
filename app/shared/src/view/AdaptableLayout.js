// @flow

import * as React from 'react';

import Device from '../Device';

type Props = {|
  renderOnWide?: React.Node,
  renderOnNarrow?: React.Node,
|};

type State = {|
  wideLayout: boolean,
|};

export default class AdaptableLayout extends React.Component<Props, State> {
  unsubscribeDimensionListener: Function = () => {};

  state = {
    wideLayout: Device.isWideLayout(),
  };

  componentDidMount = () => {
    this.unsubscribeDimensionListener = Device.subscribeToDimensionChanges(
      ({ width }) => {
        this.setState({
          wideLayout: width > Device.getWideDeviceThreshold(),
        });
      },
    );
  };

  componentWillUnmount = () => {
    this.unsubscribeDimensionListener();
  };

  render = () => {
    if (this.state.wideLayout === true && this.props.renderOnWide) {
      return this.props.renderOnWide;
    } else if (this.state.wideLayout === false && this.props.renderOnNarrow) {
      return this.props.renderOnNarrow;
    }

    // may return nothing:
    // 1. renderOnWide set but we have narrow layout
    // 2. renderOnNarrow set but we have wide layout
    return null;
  };
}
