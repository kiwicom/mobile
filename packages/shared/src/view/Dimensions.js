// @flow

import Provider from './DimensionsProvider';
import Consumer from './DimensionsConsumer';

const Dimensions = {
  Provider,
  Consumer,
};

export default Dimensions;
