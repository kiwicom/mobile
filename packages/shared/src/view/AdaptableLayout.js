// @flow

import * as React from 'react';

import Dimensions from './Dimensions';
import Device from '../Device';

type Props = {|
  +renderOnWide?: React.Node,
  +renderOnNarrow?: React.Node,
|};

// may return nothing:
// 1. renderOnWide set but we have narrow layout
// 2. renderOnNarrow set but we have wide layout
export default class AdaptableLayout extends React.Component<Props> {
  render = () => {
    return (
      <Dimensions.Consumer>
        {dimensions => {
          if (!dimensions) {
            return null;
          }
          const wideLayout = Device.isWideLayout(dimensions);
          if (wideLayout === true && this.props.renderOnWide) {
            return this.props.renderOnWide;
          } else if (wideLayout === false && this.props.renderOnNarrow) {
            return this.props.renderOnNarrow;
          }
          return null;
        }}
      </Dimensions.Consumer>
    );
  };
}
