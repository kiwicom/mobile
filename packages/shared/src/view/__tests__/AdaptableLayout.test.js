// @flow

import * as React from 'react';
import Renderer from 'react-test-renderer';

import AdaptableLayout from '../AdaptableLayout';
import Dimensions from '../Dimensions';

it('renders empty element without props', () => {
  expect(
    Renderer.create(
      <Dimensions.Provider dimensions={null}>
        <AdaptableLayout />
      </Dimensions.Provider>,
    ),
  ).toMatchSnapshot();
});

it('throw error when called without provider', () => {
  const NODE_ENV = process.env.NODE_ENV;
  process.env.NODE_ENV = 'something else than test';

  // Suppress this error:
  // Consider adding an error boundary to your tree to customize error handling behavior.
  jest.spyOn(console, 'error').mockImplementation(() => {});

  expect(() =>
    Renderer.create(<AdaptableLayout />),
  ).toThrowErrorMatchingSnapshot();

  process.env.NODE_ENV = NODE_ENV;
});
