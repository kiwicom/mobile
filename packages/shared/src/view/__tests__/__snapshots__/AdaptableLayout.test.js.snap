// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty element without props 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
